### PR TITLE
Add Interval.Intersects method

### PR DESCRIPTION
**Pull request by mattjohnsonpint**
It is sometimes useful to see if an interval overlaps with another interval.  The math is straightforward, but somewhat non-intuitive.  Thus a built-in function seems appropriate.

### CODE REVIEW

File: src/NodaTime/Interval.cs [hunk 1]
```diff
@@ -155,6 +155,14 @@ public Instant End
         [Pure]
         public bool Contains(Instant instant) => instant >= start && instant < end;
 
+        /// <summary>
+        /// Returns whether or not this interval intersects with the given interval.
+        /// </summary>
+        /// <param name="interval">Interval to test.</param>
+        /// <returns>True if this interval intersects with the given interval; false otherwise.</returns>
+        [Pure]
+        public bool Intersects(Interval interval) => interval.start < end && interval.end > start;
```
Comment (jskeet) [inline src/NodaTime/Interval.cs]: Would it be useful to change this to an `Intersection` method that returns a `Nullable<Interval>` I wonder?

**Review by jskeet — COMMENTED** [review]
Implementation is fine, but I'm uncertain about how best to expose it.

**Comment by malcolmr** [comment]
This is a good idea (and note that due to `start` and `end` not being exposed directly, it would be harder to implement in user code). However, I think it needs some work around empty intervals.

At present, two empty intervals (even those equal to one another) do not intersect. This is reasonable, especially if you consider the definition of intersect to be "contain an instant in common", as empty intervals contain no instants.

However, a non-empty interval will intersect with all empty intervals originating at [start+1, end), which seems odd (both because of the +1, and because of the inconsistency with `Contains()`). Would it make sense to filter out all empty intervals on both sides? (And, if so, to document the semantics of `Intersect` as based on the `Contains()` relation, and note what this means for empty intervals?)

Also, more test coverage around empty/half-open/infinite intervals would be nice.

File: src/NodaTime/Interval.cs (hunk 1 above)
Comment (malcolmr) [reply src/NodaTime/Interval.cs]: If you're going to add that, it's probably worth adding union and perhaps difference/symmetric difference.

**Comment by jskeet** [comment]
I was going to check for prior art in the java.time API, but that doesn't even *have* interval, interestingly. Joda Time has `overlap` and `gap`, but it's not clear what `overlap` does in the face of an empty interval.

If we change to use `Intersection` returning an `Interval?` would we want to make `nonempty.Intersection(empty)` return `empty` if the start/end time of `empty` is within `nonempty`, and null otherwise?

**Comment by jskeet** [comment]
See also #649.

**Comment by malcolmr** [comment]
@jskeet I think it would be reasonable either to define intersection to be based on "intervals contain an instant common to both" _or_ on "ranges covered by the instants overlap" (where empty intervals have a defined empty range), just as long as we document it.

**Comment by jskeet** [comment]
This is really old, but I've been thinking it would be good to clear up the repo in terms of both PRs and issues.

I think I'm most in favour of creating Intersection and Union methods over this. (And possibly Difference as well.) @mattjohnsonpint I know it's been years, but any thoughts on that?

**Comment by mattjohnsonpint** [comment]
Hey, sorry but I haven't given much thought to this project in a long while.  I'd say do whatever you think is best.
